pkg/artifactory/resource_artifactory_oauth_settings.go

```diff
@@ -0,0 +1,247 @@
+package artifactory
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"gopkg.in/yaml.v2"
+)
+
+type OauthSecurity struct {
+	Oauth OauthSettingsWrapper `yaml:"security"`
+}
+
+type OauthSettingsWrapper struct {
+	Settings OauthSettings `yaml:"oauthSettings"`
+}
+
+type OauthSettings struct {
+	EnableIntegration        bool                             `yaml:"enableIntegration" json:"enabled"`
+	PersistUsers             bool                             `yaml:"persistUsers" json:"persistUsers"`
+	AllowUserToAccessProfile bool                             `yaml:"allowUserToAccessProfile" json:"allowUserToAccessProfile"`
+	OauthProvidersSettings   map[string]OauthProviderSettings `yaml:"oauthProvidersSettings"`
+	AvailableTypes           []OauthType                      `json:"availableTypes"`
```

Comment:
??? Wait... I am baffled. Some should marshal as JSON and some as YAML? Or, is this json jammed into a yaml field?

Reply:
Struct fields with the `yaml` tag are used to marshal as YAML for `PATCH` requests
Struct fields with the `json` tag are used to unmarshal from the undocumented API endpoints JSON response

All fields from the undocumented API endpoints are included, so there are extras (e.g., `AvailableTypes`)

This particular field, `AvailableTypes`, could be used to validate that only a supported OAuth provider types are set in `OauthProvidersSettings`.  That validation is not currently implemented, but could certainly be added.

Reply:
What are you proposing for validation? Static validation or dynamic?